File: cpp-implementation/src/plotter.cpp
#include "plotter.hpp"

namespace plt = matplotlibcpp;

void plot_data(
    const std::vector<double>& time,
    const std::vector<double>& true_data,
    const std::vector<double>& measured_data,
    const std::vector<double>& estimated_data,
    const std::string& data_type
) {
    std::string title, true_label, measured_label, estimated_label, save_name;

    if (data_type == "displacement") {
        title = "Displacement: True vs Measured vs Estimated";
        true_label = "True Displacement";
        measured_label = "Measured Displacement";
        estimated_label = "Estimated Displacement";
        save_name = "displacement_plot.png";
    } else if (data_type == "velocity") {
        title = "Velocity: True vs Measured vs Estimated";
        true_label = "True Velocity";
        measured_label = "Measured Velocity";
        estimated_label = "Estimated Velocity";
        save_name = "velocity_plot.png";
    } else {
        throw std::invalid_argument("Invalid data_type. Choose 'displacement' or 'velocity'.");
    }
    
    // Plot the data
    plt::named_plot(true_label, time, true_data, "r--");
    plt::named_plot(measured_label, time, measured_data, "b--");
    plt::named_plot(estimated_label, time, estimated_data, "g");
    plt::title(title);
    plt::legend();

    // Save the plot
    std::string full_save_path = "../results/" + save_name;
    plt::save(full_save_path);
    plt::show();
}

File: python-implementation/kalman_filter/plotter.py
import matplotlib.pyplot as plt
from typing import List

def plot_data(
    time: List[float],
    true_data: List[float],
    measured_data: List[float],
    estimated_data: List[float],
    data_type: str,  # 'displacement' or 'velocity'
):
    """
    A general function to plot true, measured, and estimated data for displacement or velocity.
    Arguments:
        time: List of time steps.
        true_data: True data values.
        measured_data: Measured data values.
        estimated_data: Estimated data values.
        data_type: Type of data to plot ('displacement' or 'velocity').
    """
    # Set titles and labels based on the data type
    if data_type == 'displacement':
        title = "Displacement: True vs Measured vs Estimated"
        true_label = "True Displacement"
        measured_label = "Measured Displacement"
        estimated_label = "Estimated Displacement"
    elif data_type == 'velocity':
        title = "Velocity: True vs Measured vs Estimated"
        true_label = "True Velocity"
        measured_label = "Measured Velocity"
        estimated_label = "Estimated Velocity"
    else:
        raise ValueError("Invalid data_type. Choose 'displacement' or 'velocity'.")
    
    # Plot the data
    plt.figure()
    plt.plot(time, true_data, 'r--', label=true_label)
    plt.plot(time, measured_data, 'b--', label=measured_label)
    plt.plot(time, estimated_data, 'g', label=estimated_label)
    plt.title(title)
    plt.legend()
    plt.grid(True)

    # Automatically generate the save name based on the data type
    save_name = f"results/{data_type}_plot.png"
    plt.savefig(save_name)
    plt.show()
